Игра Сапер на Базе MVC

DavidGelmanSerg opened the pull request.

DavidGelmanSerg committed:
Игра Сапер на базе MVC

DavidGelmanSerg committed:
Delete SimpleCalculator/src/exceptions directory

DavidGelmanSerg committed:
Delete SimpleCalculator/src/expression directory

DavidGelmanSerg committed:
Delete SimpleCalculator/src/output directory

DavidGelmanSerg committed:
Delete SimpleCalculator/src/InputParser.java

DavidGelmanSerg committed:
Удалены лишние файлы

Yuksovich commented on SimpleCalculator/src/Main.java:
```diff
@@ -1,67 +1,22 @@
-import exceptions.InvalidArgumentException;
```
ветку надо было отпочковывать от main - тогде не было бы этих красных строк, да и в целом так правильнее

Yuksovich commented on SimpleCalculator/src/view/ui/StatePanel.java:
```diff
@@ -0,0 +1,48 @@
+    }
+
+    public void setLooseSmile() {
+        ImageIcon img = new ImageIcon(Objects.requireNonNull(getClass().getResource("pictures/death.png")));
```
картинки то не добавлены в проект

DavidGelmanSerg replied:
Сделано

Yuksovich commented on SimpleCalculator/src/controller/ClassicSapperController.java:
```diff
@@ -0,0 +1,28 @@
+import common.SapperDifficulty;
+
+public class ClassicSapperController extends SapperController {
+    public ClassicSapperController() {
```
пустые конструкторы не надо писать, они генерятся компилятором, если других нет

DavidGelmanSerg replied:
Исправил

Yuksovich commented on SimpleCalculator/src/common/SapperDifficulty.java:
```diff
@@ -0,0 +1,7 @@
+package common;
+
+public enum SapperDifficulty {
+    BEGINNER_DIFFICULTY,
```
в названии енума уже есть Difficulty, дублировать не надо

DavidGelmanSerg replied:
Сделано

Yuksovich commented on SimpleCalculator/src/common/CellData.java:
```diff
@@ -0,0 +1,31 @@
+
+    public CellData(int row, int col, int bombsAround, CellState state) {
+        this.row = row;
+        column = col;
```
лучше придерживаться единого стиля и тут тоже писать this.column = column, тем более такой конструктор генерирует идея

DavidGelmanSerg replied:
Исправил

Yuksovich commented on SimpleCalculator/src/controller/SapperControllerFactory.java:
```diff
@@ -0,0 +1,9 @@
+package controller;
+
+public class SapperControllerFactory {
```
тут стоит сделать приватный конструктор, чтобы никто не додумался создать экземпляр данного класса

DavidGelmanSerg replied:
Удалил фабрику.

Yuksovich commented on SimpleCalculator/src/controller/SapperType.java:
```diff
@@ -0,0 +1,5 @@
+package controller;
+
+public enum SapperType {
+    CLASSIC
```
а какие еще предполагаются?

DavidGelmanSerg replied:
Вообще, когда я читал про типы саперов, то там были и 3D варианты игры. Ну я и подумал, что для большей гибкости Абстрактный класс сделать. Но В итоге оказалось, что лучше просто другой контроллер написать, если уж на то пошло.
Поэтому удалил

Yuksovich commented on SimpleCalculator/src/controller/SapperController.java:
```diff
@@ -0,0 +1,23 @@
+import common.SapperDifficulty;
+import model.SapperModel;
+
+public abstract class SapperController {
```
для чего нужен абстрактный контроллер?

Yuksovich replied:
больше похоже, что тут был бы правильнее интерфейс, хотя и он кажется не сильно то нужен

DavidGelmanSerg replied:
Исправил. Свел все к единственному контроллеру

Yuksovich commented on SimpleCalculator/src/model/SapperModel.java:
```diff
@@ -0,0 +1,92 @@
+
+import java.beans.PropertyChangeEvent;
+import java.beans.PropertyChangeListener;
+import java.beans.PropertyChangeSupport;
```
вообще не советую использовать java.beans - это старая страшная штука, которая уже не используется

DavidGelmanSerg replied:
Исправил: Создал свою реализацию паттерна наблюдатель.

Yuksovich commented on SimpleCalculator/src/model/SapperModel.java:
```diff
@@ -0,0 +1,92 @@
+    }
+
+    private boolean isStarted() {
+        return status.equals(SapperGameStatus.STARTED);
```
енумы надо сравнить оператором ==, идея должна была подсказать это. Суть в том, что енум по факту всегда в одном экземпляре, поэтому их корректно сравнивать через ==, который проверяет является ли это одним и тем же объектом. При этом equals тоже даст корректный результат, но == заодно осуществляет проверку типов, на уровне компиляции, т.е. нельзя сравнить String и enum, компилятор ругнется, при этом через equal можно

DavidGelmanSerg replied:
Сделано

Yuksovich commented on SimpleCalculator/src/model/SapperModel.java:
```diff
@@ -0,0 +1,92 @@
+    }
+
+    private void fireWin() {
+        support.firePropertyChange(new PropertyChangeEvent(this, "", SapperEventType.WIN, null));
```
ввиду того, что ты использовал данный компонент, у тебя появляются непонятные параметры в методе, неговоря уже о том, что надо передать this. Т.е. вся модель поедет во view

DavidGelmanSerg replied:
Исправил: Создал свою реализацию паттерна наблюдатель.

Yuksovich commented on SimpleCalculator/src/model/field/SapperField.java:
```diff
@@ -0,0 +1,224 @@
+        List<FieldCell> cellsWithBombs = new ArrayList<>();
+        while (bombsCount != bombs) {
+            int row = random.nextInt(0, side);
+            int column = random.nextInt(0, side);
```
не очень хороший алгоритм расстановки бомб, слишком высокая вероятность коллизий, особенно если добавить кастомное поле (в оригинальной игре было такое) с количеством мин, почти равным размеру поля - можно прям зависнуть на расстановке, хотелось бы увидеть более интересный алгоритм

DavidGelmanSerg replied:
Исправил. Теперь алгоритм умеет отслеживать ячейки без бомб, что не должно приводить к коллизиям.

Yuksovich commented on SimpleCalculator/src/model/field/SapperField.java:
```diff
@@ -0,0 +1,224 @@
+        return getCellData(cells[x][y]);
+    }
+
+    public CellData[] getCellsWhenLoose(int x, int y) {
```
зачем оперировать массивом, когда есть списки, например?

Yuksovich replied:
в плане двумерного массива для поля - это нормально, но здесь кажется неудобно

DavidGelmanSerg replied:
Исправил. свел все к оперированию списками.

Yuksovich commented on SimpleCalculator/src/view/ui/SapperGameFrame.java:
```diff
@@ -0,0 +1,91 @@
+                field.update(cellsWithBombs);
+                field.disableField();
+                statePanel.setLooseSmile();
+                JOptionPane.showMessageDialog(this, "You loose :(");
```
стоит фразы воводимые на экран вынести в отдельное место, чтобы проще было исправить если что, вообще в джава есть messages API, тогда можно будет легко поддержать и локализацию

DavidGelmanSerg replied:
а что такое Messages API? это класс Locale?
я нашел описание JMS (Java Message Service) в сети, но не думаю, что это то, что нужно.
Можешь, пожалуйста, подсказать, куда нужно копать, чтобы я это посмотрел?

Yuksovich commented on SimpleCalculator/src/view/ui/SweeperMenu.java:
```diff
@@ -0,0 +1,49 @@
+    private SapperController controller;
+
+    public SweeperMenu() {
+        JMenu menu = new JMenu("Меню");
```
а тут русский

DavidGelmanSerg replied:
Исправил: добавил поддержку локализации с сохранением выбранных настроек

Yuksovich commented on SimpleCalculator/src/view/ui/field/FieldPanel.java:
```diff
@@ -0,0 +1,67 @@
+        setVisible(true);
+    }
+
+    public FieldPanel() {
```
для чего это тут?

DavidGelmanSerg replied:
Исправил

Yuksovich commented on SimpleCalculator/src/view/ui/SapperGameFrame.java:
```diff
@@ -0,0 +1,91 @@
+
+    @Override
+    public void propertyChange(PropertyChangeEvent evt) {
+        var eventType = (SapperEventType) evt.getOldValue();
```
вот стоило написать свои ивенты и свои листенеры, тогда не было бы этих кастов. Кастование вообще стоит избегать, оно нарушает строгую типизацию и могут возникнуть внезапные ошибки, потому как компилятор не может их проверить

DavidGelmanSerg replied:
Исправил: Создал свою реализацию паттерна наблюдатель.

DavidGelmanSerg committed:
Добавлены картинки в проект. Перемещены файлы FieldPanel и CellButton в общий пакет view.ui

DavidGelmanSerg committed:
View:;
1) Добавлена поддержка локализации и сохранение настроек;
2) Добавлена реализация однопоточной Шины событий с ленивой инициализацией;
3) SweeperMenu переименован в SapperMenu для поддержки концепта именований;
4) В SapperMenu больше не передается контроллер -> Меню работает через шину событий;
5) FieldPanel оперирует списком Объектов CellButton вместо обычного массива;
6) SapperGameFrame Полностью переведен на шину событий SapperViewEventBus;
7) SapperGameFrame Полностью переведен на обработку событий
SapperEvent вместо событий PropertyChangeEvent;
8) В проект добавлены Ассеты;

Controller:
1) Контроллер сведен от фабрики к единственному классу;

Model:
1) в класс FieldCell добавлен метод equalsCoordinates для поиска ячейки поля в списке ячеек по координатам;
2) В методах проверки состояния проверка equals() заменена на оператор равенства;
3) Добавлен новый тип сложности SapperDifficulty - CURRENT и доработана фабрика полей;
4) Изменена функция getCellData в классе FieldCell для исключения циклической зависимости с классом CellData;
5) В классе SapperField ячейки FieldCell хранятся в списке, а не массиве;
6) Добавлены вспомогательные методы для работы со списком ячеек;
7) Доработан алгоритм расстановки бомб - теперь он учитывает ячейки, в которых бомба была проставлена и не повторяется;
8) класс модели полностью переведен на собственную реализацию паттерна наблюдатель;

DavidGelmanSerg committed:
Merge remote-tracking branch 'origin/Sapper' into Sapper

DavidGelmanSerg committed:
Добавлен Таймер и таблица рекордов с возможностью хранения в H2 БД; Добавлены новые пакеты ресурсов; Переработан SapperViewEvent;

DavidGelmanSerg committed:
Добавлен Таймер и таблица рекордов с возможностью хранения в H2 БД; Добавлены новые пакеты ресурсов; Переработан SapperViewEvent;

DavidGelmanSerg committed:
Проект переведен на систему сборки Gradle. Изменена структура проекта, соответствующая структуре с одним корневым проектом и множеством отдельных подпроектов - модулей.

DavidGelmanSerg committed:
Исправлены 2 бага проекта sapper:
1. Множественное нажатие на ячейку для ее открытия
2. Определение победы

Добавлено логирование в модуль sapper (slf4j + logback) с логированием в файл и консоль.
